docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,347 @@
+      - ['MinChange' stabilisation prevents response to node CPU exhaustion](#minchange-stabilisation-prevents-response-to-node-cpu-exhaustion)
+      - [HPA-VPA policy fusion model does not account for multiplicative interaction](#hpa-vpa-policy-fusion-model-does-not-account-for-multiplicative-interaction)
+    - [Other Benefits to Replacing the Existing Solution](#other-benefits-to-replacing-the-existing-solution)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Rationale](#rationale)
+    - [Design Outline](#design-outline)
+    - [Element: Gardener Custom Metrics Provider Component](#element-gardener-custom-metrics-provider-component)
+    - [Element: New Custom Pod Metric for ShootKapis](#element-new-custom-pod-metric-for-shootkapis)
+    - [Element: HPA](#element-hpa)
+    - [Element: VPA](#element-vpa)
+    - [Transition Strategy](#transition-strategy)
+  - [Discussion and Limitations](#discussion-and-limitations)
+      - [Gardener-custom metrics precludes the use of other sources of custom metrics](#gardener-custom-metrics-precludes-the-use-of-other-sources-of-custom-metrics)
+      - [Observed fluctuations in actual compute efficiency affecting ShootKapi workloads](#observed-fluctuations-in-actual-compute-efficiency-affecting-shootkapi-workloads)
+  - [Alternatives](#alternatives)
+  - [References](#references)
+
+## Summary
+When it comes to autoscaling shoot control plane `kube-apiserver` instances (ShootKapi hereafter), Gardener needs
+both stability and efficiency (accurate scaling). The existing approach of fusing HPA and VPA into the 2-dimensional
+autoscaler which is HVPA, poses severe algorithmic limitations which manifest as stability and efficiency issues in the
+field. 
+
+This document outlines the main deficiencies in the existing solution and proposes a simple design which overcomes them,
+reduces maintenance effort by reducing Gardener's custom code base, and enables subsequent reduction in ShootKapi
+compute cost.
+
+## Motivation
+Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances (ShootKapi).
+This need is a consequence of the following concerns:
+1. Gardener supplies K8s clusters of highly diverse compute scale and utilization - ranging from a couple of small
+   practically idle nodes, to hundreds of heavily utilized giants. Furthermore, the ratio of 'cluster compute capacity'
+   to 'ShootKapi load' substantially varies between different applications utilizing Gardener shoots. As a result,
+   compute resources required by ShootKapis of different shoots vary by two orders of magnitude.
+2. The compute cost of ShootKapi workloads constitutes a major part of Gardener's overall compute cost. Small relative
+   waste in sizing ShootKapis results in considerable absolute cost.
+3. Kube-apiserver does not scale well horizontally, in a sense that extra replicas come with substantial compute overhead.   
+
```

Comment:
The main disadvantages I am aware of are (missing in the motivation here - at least for me):
- HVPA scales first horizontally (up to a magic number of 4) before it scales vertically (that's how it's used in practice), which adds more wasteful replicas than actually necessary (every replica needs to have its caches, etc.)
- The number 4 is somewhat random (results from 1 performance test (!); assumptions that I have strong doubts about; disproven by some Gardener hosters that run seeds with 6 KAPIs and possibly even more). In addition, it fits poorly to HA (one zone running twice as many as the other two zones (in a scenario with three zones)).
- Whenever HVPA makes a decision, the entire deployment is rolled, resulting in constant waste. While @voelzmo notes that VPA will "roll a pod" at the latest after 12h anyhow, the current HVPA implementation seems even more wasteful as "all pods are rolled" at once and all need to start from scratch to fill their caches, etc.
- Even if it would be smarter, it lacks proper hysteresis support and may/will jiggle.

Reply:
>HVPA scales first horizontally

True, but if I only state that HVPA has such limitation, I'd be formally incorrect. Even the very first example in [HVPA's readme](https://github.com/gardener/hvpa-controller/blob/master/README.md#spec) suggests the opposite. I find myself in a position where I cannot make a publication regarding HVPA, which contradicts HVPA's official documentation, and not provide justification for my claims. That led me to the need to somehow reference HVPA's algorithmic limitations.

My initial call (right or wrong) was to omit from the GEP the primary HVPA issues you listed, because they are not fundamental in nature, but rather a consequence from the algorithmic limitations I listed instead. **However, I take your point that the primary practical issues are important and belong in the document, and I'll add them.**
<br/>
>Even if it would be smarter, it lacks proper hysteresis support and may/will jiggle.

I think this is covered, albeit with a different wording (let me know if you think differently after seeing the following):
>the resulting autoscaler lacks joint hysteresis behavior (...) risk of oscillating on closed contours
_(in section "Algorithmic limitation: fusing the behavior of two blackbox systems")_

Reply:
Yes, mixed mode is in the docs of HVPA, but practically, it wouldn't work well, which is why it first scales horizontally, than vertically for everybody in real life. Anyway, it was a comment and I am fine however you carry on.

Reply:
A list of the primary operational issues resulting from the use of HVPA is now available in the GEP.